File: aat-lib/src/main/java/ch/bailu/aat_lib/dispatcher/CurrentLocationSource.java
package ch.bailu.aat_lib.dispatcher;


import ch.bailu.aat_lib.gpx.GpxInformation;
import ch.bailu.aat_lib.gpx.InfoID;
import ch.bailu.aat_lib.service.ServicesInterface;

public class CurrentLocationSource extends ContentSource {

    private final ServicesInterface services;
    private final Broadcaster broadcaster;


    private final BroadcastReceiver onLocationChange = new BroadcastReceiver () {
        @Override
        public void onReceive(Object ...o) {
            sendUpdate(InfoID.LOCATION, services.getLocationService().getLocationInformation());
        }

    };



    public CurrentLocationSource(ServicesInterface services, Broadcaster broadcaster) {
        this.services = services;
        this.broadcaster = broadcaster;
    }





    @Override
    public void requestUpdate() {
        sendUpdate(InfoID.LOCATION, services.getLocationService().getLocationInformation());
    }


    @Override
    public void onPause() {
        broadcaster.unregister(onLocationChange);
    }

    @Override
    public void onResume() {
        broadcaster.register(onLocationChange, AppBroadcaster.LOCATION_CHANGED);
    }

    @Override
    public int getIID() {
        return InfoID.LOCATION;
    }

    @Override
    public GpxInformation getInfo() {
        return services.getLocationService().getLocationInformation();
    }

}

File: aat-lib/src/main/java/ch/bailu/aat_lib/dispatcher/CurrentLocationSource.kt
package ch.bailu.aat_lib.dispatcher

import ch.bailu.aat_lib.broadcaster.AppBroadcaster
import ch.bailu.aat_lib.broadcaster.BroadcastReceiver
import ch.bailu.aat_lib.broadcaster.Broadcaster
import ch.bailu.aat_lib.gpx.GpxInformation
import ch.bailu.aat_lib.gpx.InfoID
import ch.bailu.aat_lib.service.ServicesInterface

class CurrentLocationSource(
    private val services: ServicesInterface,
    private val broadcaster: Broadcaster
) : ContentSource() {
    private val onLocationChange = BroadcastReceiver {
        sendUpdate(
            InfoID.LOCATION,
            services.locationService.getLocationInformation()
        )
    }

    override fun requestUpdate() {
        sendUpdate(InfoID.LOCATION, services.locationService.getLocationInformation())
    }

    override fun onPause() {
        broadcaster.unregister(onLocationChange)
    }

    override fun onResume() {
        broadcaster.register(AppBroadcaster.LOCATION_CHANGED, onLocationChange)
    }

    override fun getIID(): Int {
        return InfoID.LOCATION
    }

    override fun getInfo(): GpxInformation {
        return services.locationService.getLocationInformation()
    }
}
